### PR TITLE
Add Cadence best practice docs.

### PR DESCRIPTION
**Pull request by 10thfloor**
## Description
Adds @robmyers Cadence best practice docs to Cadence section.
https://www.notion.so/dapperlabs/Flow-and-Cadence-Development-Best-Practice-874f51cf092e4a37bf2aaaced6e1e4de

### CODE REVIEW

File: docs/content/guides/cadence/dragons.mdx [hunk 1]
```diff
@@ -0,0 +1,117 @@
+---
+title: Cadence Dragons
+---
+
+This is an opinionated list of issues that can be improved if they are found in Cadence code intended for production.
+
+## Security And Robustness
```
Comment (psiemens) [inline docs/content/guides/cadence/dragons.mdx:7]:
```suggestion
## Security and Robustness
```

File: docs/content/guides/cadence/dragons.mdx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+A public function on a contract that creates a resource can be called by any account.
+
+If that resource has functions that emit events, any account can create an instance of that resource and emit those events.
+
+If those events are meant to indicate actions taken using a single instance of that resource (for example an admin object, switchboard, or registry), or instances created through a particular workflow, then this will make event log search and management harder because events from other instances may be mixed in with the ones that you wish to examine.
+
+To fix this, if there should be only a single instance of the resource it should be created and `link()`ed to a public path in an admin account's storage during the contracts's initializer.
+
+### Named Value Fields are Preferable
```
Comment (psiemens) [inline docs/content/guides/cadence/dragons.mdx:19]:
```suggestion
### Named Value Fields Are Preferable
```

File: docs/content/guides/cadence/dragons.mdx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+Where contracts, resources, and scripts all have to repeatedly refer to the same constant values, entering these manually is a potential source of error. These are known as [magic values](<https://en.wikipedia.org/wiki/Magic_number_(programming)>).
+
+Rather than typing the values manually in functions, transactions and scripts, store the values once in fields of the contract responsible and then access them via those fields. This is the [Named Value Field pattern](../design-patterns/#named-value-field)
+
+### Public Functions And Fields Should Be Avoided
```
Comment (psiemens) [inline docs/content/guides/cadence/dragons.mdx:25]:
```suggestion
### Public Functions and Fields Should Be Avoided
```

File: docs/content/guides/cadence/dragons.mdx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+Be sure to keep track of access modifiers when structuring your code, and make public only what should be public.
+
+### Array or Dictionary Fields should be private
```
Comment (psiemens) [inline docs/content/guides/cadence/dragons.mdx:29]:
```suggestion
### Array or Dictionary Fields Should Be Private
```

File: docs/content/guides/cadence/dragons.mdx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+### Problem
```
Comment (psiemens) [inline docs/content/guides/cadence/dragons.mdx:31]:
```suggestion
#### Problem
```

File: docs/content/guides/cadence/dragons.mdx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+import Array from 0x01
+
+transaction {
+    execute {
+		    Array.shouldbeConstantArray[0] = 1000
+    }
+}
+```
+
+### Solution
```
Comment (psiemens) [inline docs/content/guides/cadence/dragons.mdx:58]:
```suggestion
#### Solution
```

File: docs/content/guides/cadence/dragons.mdx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+Note that this doesn't mean that the restricted type should _always_ be emitted, only if it's specifically `AnyStruct` or `AnyResource` .
+
+A good example of when the code should specify the type being restricted is checking the FLOW balance: the code must borrow `&FlowToken.Vault{FungibleToken.Balance}`, in order to ensure that it gets a FLOW token balance, and not just `&{FungibleToken.Balance}`, any balance – the user could store another object that conforms to the balance interface and return whatever value as the amount.
+
+### Plural Names For Arrays And Maps Are Preferable
+
+e.g. `accounts` rather than `account` for an array of accounts.
+
+This signals that the field or variable is not scalar. It also makes using the singular when iterating the field clearer.
```
Comment (psiemens) [inline docs/content/guides/cadence/dragons.mdx:117]:
```suggestion
This signals that the field or variable is not scalar. It also makes it easier to the singular form for a variable name during iteration.
```

File: docs/content/guides/cadence/dragons.mdx [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+`account` / `accounts` is better than `array` / `element`.
+
+`providerAccount` / `tokenRecipientAccount` is better than `acct1` / `acct2`.
+
+Unless of course you are dealing with entities that really are best named that way, for example in utility code.
+
+### Omitting "Any" Types In Constraints Is Preferable
+
+Prefer `&{Receiver}` to `&AnyResource{Receiver}` .
```
Comment (psiemens) [inline docs/content/guides/cadence/dragons.mdx:105]:
```suggestion
Prefer `&{Receiver}` to `&AnyResource{Receiver}`.
```

File: docs/content/guides/cadence/dragons.mdx [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+### Public Admin Resource Creation Functions Are Unsafe
+
+This is a specific case of "Public Functions And Fields Should Be Avoided", above.
+
+A public function on a contract that creates a resource can be called by any account.
+
+If that resource provides access to admin functions then the function should not be public.
+
+To fix this, a single instance of that resource created using it then `link()`ed to a private path in an admin account's storage during the contract's initializer. If the code to create the resource is complex enough that it needs its own function, the function that creates the resource should be access(contract).
```
Comment (psiemens) [inline docs/content/guides/cadence/dragons.mdx:77]:
```suggestion
To fix this, a single instance of that resource created using it then `link()`ed to a private path in an admin account's storage during the contract's initializer. If the code to create the resource is complex enough that it needs its own function, the function that creates the resource should use `access(contract)` as its access modifier.
```

File: docs/content/guides/cadence/dragons.mdx (hunk 3 above)
Comment (10thfloor) [reply docs/content/guides/cadence/dragons.mdx:25]: Are we Chicago style or AP? lol

**Review by psiemens — APPROVED** [review]

**Pull request merged**